### PR TITLE
Add power method to truncated SVD solver

### LINKED ISSUE #346: Power method tSVD
Investigate and, if seems ok, implement a different algo for tSVD which utilizes more GPU.

### PR DESCRIPTION
**Pull request by navdeep-G**
Resolves #346

### CODE REVIEW

**Comment by mdymczyk** [comment]
Looks good, can we get some benchmarks before we merge this in? Vs. scikit and versus our old implementation. Also the memory requirements.

**Comment by navdeep-G** [comment]
@mdymczyk Sure, but I want to check the impl with @RAMitchell first.

**Comment by navdeep-G** [comment]
@mdymczyk @RAMitchell 
Initial benchmarks:

cusolverpower
cusolverpowerrandomized
cusolverpowersklearnrandomized
powerrandomized
powersklearnrandomizedarpack

**Comment by navdeep-G** [comment]
@mdymczyk @RAMitchell Updated benchmarks
Show an optimized power method (commit: https://github.com/h2oai/h2o4gpu/pull/361/commits/0b672e8304fb2cb50d4b2e191474b6f53acdf3d2) vs orignal impl. These benchmarks show the avg runtime across 5 runs.
power_avg_5runs

**Comment by navdeep-G** [comment]
Updated benchmarks with changes from https://github.com/h2oai/h2o4gpu/pull/361/commits/0b672e8304fb2cb50d4b2e191474b6f53acdf3d2
power
powervsoptimizedpowervscusolver

**Comment by mdymczyk** [comment]
@navdeep-G interesting, so it seems that the power method is still slower for x10 and x100 matrices? Have you ran the nvidia profiles to check which parts of the code are being executed the longest and can we do something about it?

File: src/gpu/data/matrix.cuh [hunk 1]
```diff
@@ -409,6 +410,17 @@ namespace tsvd
...
 	void normalize_columns(Matrix<tsvd_float>& M, DeviceContext& context);
 
+	/*
+	 * Add docs...
+	 */
+	void normalize_vector_cublas(Matrix<tsvd_float>& M, DeviceContext& context);
+
+
+	/*
+	 * Add docs...
```
Comment (RAMitchell) [inline src/gpu/data/matrix.cuh]: Add documentation

File: src/gpu/tsvd/tsvd.cu [hunk 2]
```diff
@@ -227,6 +229,195 @@ void calculate_u(const Matrix<float> &X, const Matrix<float> &Q, const Matrix<fl
...
+	Matrix<float>A(M.rows(), M.columns());
+
+	//Create context
+	DeviceContext context;
+
+	//Multiply X and Xt and output result to XtX
+	multiply(X, X, M, context, true, false, 1.0f);
+
+//	//Temporary matrix whcih will be updated in power method iteratively
+//    Matrix<float>M(XtX.rows(), XtX.columns());
```
Comment (RAMitchell) [inline src/gpu/tsvd/tsvd.cu]: Try to avoid leaving commented out code

File: src/gpu/tsvd/tsvd.cu [hunk 3]
```diff
@@ -227,6 +229,195 @@ void calculate_u(const Matrix<float> &X, const Matrix<float> &Q, const Matrix<fl
...
+//	M.copy(XtX);
+
+	//Set up Q (V^T) and w (singular value) matrices (w is a matrix of size Q.rows() by 1; really just a vector
+	Matrix<float>Q(M.rows(), _param.k);
+	Matrix<float>w(_param.k, 1);
+	std::vector<double> w_temp(w.size());
+
```
Comment (RAMitchell) [inline src/gpu/tsvd/tsvd.cu]: Nice documentation below. This will really help the next person who reads this.

File: src/gpu/tsvd/tsvd.cu [hunk 4]
```diff
@@ -227,6 +229,195 @@ void calculate_u(const Matrix<float> &X, const Matrix<float> &Q, const Matrix<fl
...
+	 * rewrite the matrix in terms of this basis and restrict to the orthogonal space of the known eigenvector.
+	 * This is comprised in the method of deflation:
+	 *     You first find the first eigenvector v1 (with a maximal λ1), by iterating xn+1=Axn/|Axn|with a "random" initial x0. Once you have found a good approximation
+	 *     for v1, you consider B=A−λ1|v1|2 * v1vT (this simple step replaces the "rewrite in terms of this basis" above). This is a matrix that behaves like A for anything
+	 *     orthogonal to v1 and zeroes out v1. Use the power method on B again, which will reveal v2, an eigenvector of a largest eigenvalue of B.
+	 *     Then switch to C=B−λ2|v2|2 * v2vT so on.
+	 */
+	Matrix<float>b_k(_param.X_n, 1);
+	Matrix<float>b_k1(_param.X_n, 1);
+	float tol = _param.tol;
```
Comment (RAMitchell) [inline src/gpu/tsvd/tsvd.cu]: tol is an unnecessary variable. Just directly use _param.tol.

File: src/gpu/tsvd/tsvd.cu [hunk 5]
```diff
@@ -227,6 +229,195 @@ void calculate_u(const Matrix<float> &X, const Matrix<float> &Q, const Matrix<fl
...
+
+	for(int i = 0; i < _param.k; i ++){
+		//Set aside vector of randoms (n x 1)
+		b_k.random(i);
```
Comment (RAMitchell) [inline src/gpu/tsvd/tsvd.cu]: You will eventually need to add a random seed parameter to your implementation although I don't think this is required for this PR.

File: src/gpu/data/matrix.cuh (hunk 1 above)
Comment (navdeep-G) [reply src/gpu/data/matrix.cuh]: Will do.

File: src/gpu/tsvd/tsvd.cu (hunk 2 above)
Comment (navdeep-G) [reply src/gpu/tsvd/tsvd.cu]: Will remove.

File: src/gpu/tsvd/tsvd.cu (hunk 4 above)
Comment (navdeep-G) [reply src/gpu/tsvd/tsvd.cu]: Yep, not needed.

File: src/gpu/tsvd/tsvd.cu (hunk 5 above)
Comment (navdeep-G) [reply src/gpu/tsvd/tsvd.cu]: Yep, not needed for this PR but maybe I can add it.

**Pull request merged**